spyder_kernels/comms/commbase.py

```diff
@@ -0,0 +1,407 @@
+
+        spyder_msg_type = msg_dict['spyder_msg_type']
+
+        if spyder_msg_type in self._message_handlers:
+            self._message_handlers[spyder_msg_type](
+                msg_dict, buffer, load_exception)
+            return
+
+        logger.debug("No such spyder message type: %s" % spyder_msg_type)
+
+    def _handle_remote_call(self, msg, buffer, load_exception):
+        """Handle a remote call."""
+        msg_dict = msg['content']
+        if load_exception:
+            logger.debug(
+                "Exception in cloudpickle.loads : %s" % str(load_exception))
+            return
+        try:
+            return_value = self._remote_callback(
+                    msg_dict['call_name'],
+                    buffer['call_args'],
+                    buffer['call_kwargs'])
+            self._set_call_return_value(msg_dict, return_value)
+        except Exception as e:
+            tb = traceback.extract_tb(sys.exc_info()[2])
+            self._set_call_return_value(msg_dict, e, traceback=tb)
+
+    def _remote_callback(self, call_name, call_args, call_kwargs):
+        """Call the callback function for the remote call."""
+        if call_name in self._remote_call_handlers:
+            return self._remote_call_handlers[call_name](
+                *call_args, **call_kwargs)
+
+        raise CommError("No such spyder call type: %s" % call_name)
+
+    def _set_call_return_value(self, call_dict, data, traceback=None):
+        """
+        A remote call has just been processed.
+
+        This will reply if settings['blocking'] == True
+        """
+        settings = call_dict['settings']
+        if 'blocking' not in settings or not settings['blocking']:
+            return
+        content = {
+            'is_error': traceback is not None,
+            'call_id': call_dict['call_id'],
+            'call_name': call_dict['call_name']
+        }
+        if traceback is not None:
+            data = [data, traceback]
+
+        self._send_message('blocking_call_reply', content=content, data=data,
+                           comm_id=self.calling_comm_id)
+
+    def _get_call_return_value(self, call_dict):
+        """
+        A remote call has just been sent.
+
+        If settings['blocking'] == True, this will wait for a reply and return
+        the replied value.
+        """
+        settings = call_dict['settings']
+
+        if 'blocking' not in settings or not settings['blocking']:
+            return
+
+        if 'timeout' in settings:
+            timeout = settings['timeout']
+        else:
+            timeout = 3  # Seconds
+
+        call_id = call_dict['call_id']
+        call_name = call_dict['call_name']
+
+        self._wait_reply(call_id, call_name, timeout)
+
+        reply = self._call_reply_dict[call_id]
+        self._call_reply_dict.pop(call_id)
+
+        if reply['is_error']:
+            error, tb = reply['value']
+            traceback.print_list(tb)
+            raise error
```

Comment:
Do we need to raise an error here? Can't we log it instead?

Reply:
Well maybe the calling function want to be able to catch that exception and do something with it. For example, get_value would raise a KeyError if the variable is not defines.

Reply:
That example makes no sense to me. `get_value` is triggered here by doing a double-click on the Variable Explorer. So if a KeyError is generated here, it means that we're not even displaying the current variables in the kernel namespace correctly.

However, the most important question is: what would users gain by seeing this kind of errors?

Reply:
Well the user won't have to see these errors if they are handled correctly by the calling code. It the user sees this error, it means: something went wrong and the code did not handle it. This is true of any error, not just here. Otherwise why don't we log all the errors instead of having a pop-up windows showing them to the user?

Reply:
Ok, if errors are not ignored here (which I find easier), then they need to ignored on the frontend side, except for `get_value`.

Reply:
Agreed

Reply:
> then they need to ignored on the frontend side

Sorry, scratch ~~frontend~~ above. It's really "the other side" (I forgot this is placed in CommBase).

> Agreed

Ok, I agree with this too. Please take care of correctly handling errors in the handlers present here and those in spyder-ide/spyder#9343.